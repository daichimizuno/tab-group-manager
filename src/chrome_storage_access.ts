export default class ChromeStorageAccess {

  getAllTabGroup = (): Promise<string[]> => {
    return new Promise((resolve, reject) => {
      chrome.storage.sync.get(["tabGroup"], (value) => {
        if (chrome.runtime.lastError) {
          reject(chrome.runtime.lastError);
        }

        resolve(value.tabGroup as string[]);
      });
    });
  };

  addNewGroup = async (tabName: string) => {
    const tabGroupData = (await this.getAllTabGroup()) as string[] || [];
    console.log(`tab group data: ${JSON.stringify(tabGroupData)}`);

    // 既存のタブグループに新しいタブグループを追加
    tabGroupData.push(tabName)

    chrome.storage.sync.set({ "tabGroup": tabGroupData }, async () => {
      console.log(`new tab group is set, name:${tabName}`);
    });
  };
}
