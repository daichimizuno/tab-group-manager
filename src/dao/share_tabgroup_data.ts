import * as fs from "fs";
import ChromeStorageAccess from "./chrome_storage_access";

export const exportTabGroupData = async () => {
  const chromeStorageAccess = new ChromeStorageAccess();
  const tabAllGroup = await chromeStorageAccess.getAllTabGroup();

  fs.writeFileSync("./file/tab-group.json", JSON.stringify(tabAllGroup));
};

export const importTabGroupData = async (path: string) => {
  const chromeStorageAccess = new ChromeStorageAccess();
  const file = fs.readFileSync(path);

  const json = file.toJSON();
  const tabGroup = JSON.parse(json);
};
