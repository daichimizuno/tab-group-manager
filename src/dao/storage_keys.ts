export namespace StorageKeys {
  export const TAB_GROUP = "tabGroup";
  export const TAB_GORUP_LAST_INDEX = "tabGroupLastIndex";
}
