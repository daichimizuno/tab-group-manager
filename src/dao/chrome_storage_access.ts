import { assertIsDefined, AssertIsDefinedError } from "../utils/validations";
import { StorageKeys } from "./storage_keys";

export interface TabGroup {
  id: number;
  tabGroupName: string;
  tabColor: Color;
  urls: string[];
}

const Color = {
  Grey: "grey",
  Blue: "blue",
  Red: "red",
  Yellow: "yellow",
  Green: "green",
  Pink: "pink",
  Purple: "purple",
  Cyan: "cyan",
} as const;
export type Color = typeof Color[keyof typeof Color];

export default class ChromeStorageAccess {
  addNewTabGroup = async (
    tabName: string,
    tabColor: Color
  ): Promise<Boolean> => {
    const tabGroupData = ((await this.getAllTabGroup()) as TabGroup[]) || [];

    // 既に他に同じ名前のタブグループ名がある場合は追加しない
    const hasTheSameGroupNameData =
      tabGroupData.findIndex(
        (tabGroup) => tabGroup.tabGroupName === tabName
      ) !== -1;
    if (hasTheSameGroupNameData) {
      return Promise.resolve(false);
    } else {
      const tabGroupLastIndex =
        ((await this.syncUpdateLastIndex()) as number) || 1;
      const newTabGroup = {
        id: tabGroupLastIndex,
        tabGroupName: tabName,
        tabColor: tabColor,
        urls: [],
      } as TabGroup;

      tabGroupData.push(newTabGroup);

      await chrome.storage.sync.set({ tabGroup: tabGroupData }, async () => {});

      return Promise.resolve(true);
    }
  };

  addUrlToTabGroup = async (url: string, tabId: number) => {
    const tabGroupData = ((await this.getAllTabGroup()) as TabGroup[]) || [];
    if (tabGroupData.length > 0) {
      const tabGroup = tabGroupData.find((tabGroup) => tabGroup.id === tabId);
      const tabGroupIndex = tabGroupData.findIndex(
        (tabGroup) => tabGroup.id === tabId
      );

      try {
        assertIsDefined(tabGroup);

        // 既存のTabGroupにurlを追加する
        tabGroup.urls.push(url);

        // TabGroupのリストを新しいものに入れ替える
        tabGroupData.splice(tabGroupIndex, 1, tabGroup);

        await chrome.storage.sync.set(
          { tabGroup: tabGroupData },
          async () => {}
        );
      } catch (e) {
        if (e instanceof AssertIsDefinedError) {
          throw new AssertIsDefinedError("不明なエラー");
        }
      }
    }
  };

  deleteTabGroup = async (tabNames: string[]) => {
    const tabGroupData = ((await this.getAllTabGroup()) as TabGroup[]) || [];

    tabNames.forEach((tabName) => {
      const targetTabGroupIndex = tabGroupData.findIndex(
        (tabGroup) => tabGroup.tabGroupName === tabName
      );

      if (targetTabGroupIndex !== -1) {
        tabGroupData.splice(targetTabGroupIndex, 1);
      }
    });

    await chrome.storage.sync.set({ tabGroup: tabGroupData }, async () => {});
  };

  syncUpdateLastIndex = async (): Promise<number> => {
    let index = await this.getTabGroupLastIndex();
    if (isNaN(index)) {
      index = 1;
    } else {
      index = index + 1;
    }

    await this.setTabGroupLastIndex(index.toString());
    return index;
  };

  getTabGroupLastIndex = (): Promise<number> => {
    return new Promise((resolve, reject) => {
      chrome.storage.sync.get([StorageKeys.TAB_GORUP_LAST_INDEX], (value) => {
        if (chrome.runtime.lastError) {
          throw chrome.runtime.lastError;
        }

        const index =
          value.tabGroupLastIndex === undefined
            ? NaN
            : Number(value.tabGroupLastIndex);
        resolve(index);
      });
    });
  };

  setTabGroupLastIndex = (newIndex: string): Promise<void> => {
    return new Promise((resolve, reject) => {
      chrome.storage.sync.set({ tabGroupLastIndex: newIndex }, () => {
        if (chrome.runtime.lastError) {
          throw chrome.runtime.lastError;
        }

        resolve();
      });
    });
  };

  getAllTabGroup = (): Promise<TabGroup[]> => {
    return new Promise((resolve, reject) => {
      chrome.storage.sync.get([StorageKeys.TAB_GROUP], (value) => {
        if (chrome.runtime.lastError) {
          reject(chrome.runtime.lastError);
        }

        resolve(value.tabGroup as TabGroup[]);
      });
    });
  };
}
