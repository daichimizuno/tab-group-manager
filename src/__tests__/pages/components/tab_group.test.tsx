import { render, screen } from "@testing-library/react";
import React from "react";
import renderer, { act, ReactTestRendererJSON } from "react-test-renderer";
import { TabGroup } from "../../../dao/chrome_storage_access";
import TabGroupComponent from "../../../pages/components/tab_group";

const tabGroupMock = [
  {
    id: 1,
    tabGroupName: "テスト",
    tabColor: "grey",
    urls: ["https://google.com"],
  },
  {
    id: 2,
    tabGroupName: "テスト2",
    tabColor: "red",
    urls: ["https://yahoo.co.jp"],
  },
] as TabGroup[];

let tree: ReactTestRendererJSON;

beforeEach(() => {
  tree = renderer
    .create(<TabGroupComponent tabGroup={tabGroupMock} />)
    .toJSON() as ReactTestRendererJSON;
});

afterEach(() => {});

describe("tab_groupのテスト", () => {
  describe("コンポーネントのてすと", () => {
    test("スナップショットが前回の結果と一致しているかどうか", () => {
      expect(tree).toMatchSnapshot();
    });

    test("表示されるデータがTableがMockデータと正しいこと", () => {
      act(() => {
        render(<TabGroupComponent tabGroup={tabGroupMock} />);
      });

      expect(screen.getByText("テスト")).toBeTruthy();
      expect(screen.getByText("テスト2")).toBeTruthy();
      expect(screen.getByText("grey")).toBeTruthy();
      expect(screen.getByText("red")).toBeTruthy();
      expect(screen.getByText("https://google.com")).toBeTruthy();
      expect(screen.getByText("https://yahoo.co.jp")).toBeTruthy();
    });
  });
});
