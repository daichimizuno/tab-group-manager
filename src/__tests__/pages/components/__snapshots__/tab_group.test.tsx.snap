// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tab_groupのテスト コンポーネントのてすと スナップショットが前回の結果と一致しているかどうか 1`] = `
Array [
  <p
    className="MuiTypography-root MuiTypography-body1 css-4oopdw-MuiTypography-root"
    data-testid="title"
  >
    登録されたタブグループ
  </p>,
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAccordion-root MuiAccordion-rounded MuiAccordion-gutters css-1elwnq4-MuiPaper-root-MuiAccordion-root"
  >
    <div
      aria-controls="panel1a-content"
      aria-expanded={false}
      className="MuiButtonBase-root MuiAccordionSummary-root MuiAccordionSummary-gutters css-sh22l5-MuiButtonBase-root-MuiAccordionSummary-root"
      id="panel1a-header"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="MuiAccordionSummary-content MuiAccordionSummary-contentGutters css-o4b71y-MuiAccordionSummary-content"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        >
          リスト1
        </p>
      </div>
      <div
        className="MuiAccordionSummary-expandIconWrapper css-yw020d-MuiAccordionSummary-expandIconWrapper"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ExpandMoreIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
      </div>
    </div>
    <div
      className="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-bz4dnt-MuiCollapse-root"
      style={
        Object {
          "minHeight": "0px",
        }
      }
    >
      <div
        className="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
      >
        <div
          className="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
        >
          <div
            aria-labelledby="panel1a-header"
            className="MuiAccordion-region"
            id="panel1a-content"
            role="region"
          >
            <div
              className="MuiAccordionDetails-root css-15v22id-MuiAccordionDetails-root"
            >
              <div
                className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-11xur9t-MuiPaper-root-MuiTableContainer-root"
              >
                <table
                  aria-label="simple table"
                  className="MuiTable-root css-quj9j6-MuiTable-root"
                  role={null}
                >
                  <thead
                    className="MuiTableHead-root css-15wwp11-MuiTableHead-root"
                    role={null}
                  >
                    <tr
                      className="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
                      role={null}
                    >
                      <th
                        aria-sort={null}
                        className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                        scope="col"
                      >
                        タブ名
                      </th>
                      <th
                        aria-sort={null}
                        className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1azl6jz-MuiTableCell-root"
                        scope="col"
                      >
                        タブ色
                      </th>
                      <th
                        aria-sort={null}
                        className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1azl6jz-MuiTableCell-root"
                        scope="col"
                      >
                        url
                      </th>
                    </tr>
                  </thead>
                  <tbody
                    className="MuiTableBody-root css-apqrd9-MuiTableBody-root"
                    role={null}
                  >
                    <tr
                      className="MuiTableRow-root css-34nofg-MuiTableRow-root"
                      role={null}
                    >
                      <th
                        aria-sort={null}
                        className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
                        scope="row"
                      >
                        テスト
                      </th>
                      <td
                        aria-sort={null}
                        className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
                      >
                        grey
                      </td>
                      <td
                        aria-sort={null}
                        className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
                      >
                        https://google.com
                      </td>
                    </tr>
                    <tr
                      className="MuiTableRow-root css-34nofg-MuiTableRow-root"
                      role={null}
                    >
                      <th
                        aria-sort={null}
                        className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
                        scope="row"
                      >
                        テスト2
                      </th>
                      <td
                        aria-sort={null}
                        className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
                      >
                        red
                      </td>
                      <td
                        aria-sort={null}
                        className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
                      >
                        https://yahoo.co.jp
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
